### DISCUSSION TITLE
multi objective Bayesian optimization

### DISCUSSION THREAD

**Discussion by lovecodng** [meta-pytorch/botorch#809] [Q&A] [answered]
Does this package supports multi objective Bayesian optimization (MOBO)? If so is there a documentary or examples that can be used to understand how MOBO is applied using this pakckage?

**Comment by saitcakmak** [comment, marked as answer]
This tutorial may be helpful: https://botorch.org/tutorials/multi_objective_bo

In addition, Ax, which uses BoTorch under the hood, may also be interesting: https://ax.dev/tutorials/multiobjective_optimization.html. You may prefer to use Ax if you're interested in optimizing some objective without dealing with all the details, and BoTorch if you're doing BO research / developing your own methods / need lots of customization etc.

**Reply by lovecodng** [reply]
Thank you for your answer, do you know how botorch handle the trade-off between the objectives, does it use Pareto Front?

**Reply by saitcakmak** [reply]
The short answer is yes, it uses the Pareto front. The tutorial above demonstrates qEHVI and qParEGO. qEHVI explicitly uses the hyper-volume improvement on the Pareto front, and qParEGO uses random Chebyshev scalarizations to select the next point to evaluate. In the end, you still have a set of evaluated points, for which you can calculate the Pareto front and pick what solution to implement. 

You can check out this paper for more detail on these acquisition functions:[[Daulton2020qehvi]](https://papers.nips.cc/paper/2020/hash/6fec24eac8f18ed793f5eaad3dd7977c-Abstract.html)

**Reply by lovecodng** [reply]
Thank you

**Reply by lovecodng** [reply]
@saitcakmak is it possible to use two set of multidimensional data instead of one I mean one set for each objective, For instance, if the first objective function is fitted using data x, then the second objective function is fitted using data x multiplied by constant array (c), such as x*c.

**Reply by saitcakmak** [reply]
You should be able to do that using `ModelListGP` (https://botorch.org/docs/models#model-list-of-single-task-gps). It is essentially a wrapper around multiple `SingleTaskGP` models (trained on different inputs) that acts as if it is a single multi-output model. 

If I'm not mistaken, it also works with other single-output models such as `FixedNoiseGP`, which may be preferred if you have known observation noise levels or noise-free observations.

**Reply by lovecodng** [reply]
@saitcakmak Thank you for your reply is there an example on how to implement ModelListGP. I am currently working on this tutorial https://botorch.org/tutorials/multi_objective_bo
So what changes should I make to implement ModelListGP?

**Reply by Balandat** [reply]
basically, you'll need to create a `SingleTaskGP`, one for each of the objectives, say `gp_1` and `gp_2`, and fit them as you would with any such `SingleTaskGP`. Then you can stick them into a `ModelListGP` like so:

```
from botorch.models import ModelListGP

mo_gp = ModelListGP(gp_1, gp_2)
```
You can now use the `mo_gp` like any other GP with the acquisition function.

**Reply by lovecodng** [reply]
Thank you for your reply, the code worked fine, but what type of likelihood you suggest to use, and how I can determine the mll, because when I wrote `mll = ExactMarginalLogLikelihood(model.likelihood, model)`
I get the following error `RuntimeError: Likelihood must be Gaussian for exact inference` where `model = ModelListGP(model1, model2)`

**Reply by saitcakmak** [reply]
So, you need to train `model1` and `model2` individually (i.e., construct `mll` for these and call `fit_gpytorch_model` on them separately), then construct the `ModelListGP` instance. You can't use the `ModelListGP` to train the list of models it wraps.

**Reply by lovecodng** [reply]
Thank you for your prompt response, I wrote this code, is this right, do I miss something?
`model1 = SingleTaskGP(train_x1, train_obj1, outcome_transform=Standardize(m=train_obj1.shape[-1]))
    model_qparego2 = SingleTaskGP(train_x2, train_obj2, outcome_transform=Standardize(m=train_obj2.shape[-1]))
mll1 = ExactMarginalLogLikelihood(model1.likelihood, model1)
mll2 = ExactMarginalLogLikelihood(model2.likelihood, model2)
fit_gpytorch_model(mll1)
fit_gpytorch_model(mll1)
model = ModelListGP(model1, model2)`

**Reply by saitcakmak** [reply]
I see that you're defining `model_qparego2` but later using `model_2`. That may be leading to some errors. Otherwise, this looks like a correct use of `ModelListGP`.

**Reply by A-ep93** [reply]
There are two tutorials for multi-objective MO on the website. Here are the links: [Link1](https://botorch.org/tutorials/multi_objective_bo), [link2](https://botorch.org/tutorials/constrained_multi_objective_bo)
 Why there is not use of ModelListGP in those tutorials?

**Reply by saitcakmak** [reply]
I can't really think of any particular reason other than that using a `SingleTaskGP` leads to fewer lines of code and keeps the tutorial simple. If you want to use `ModelListGP`, you can replace the `initialize_model` method of those notebooks with the following:

```
def initialize_model(train_x, train_obj):
    models = [
        SingleTaskGP(
            train_x,
            train_obj[..., i : i + 1],
            outcome_transform=Standardize(m=1),
        )
        for i in range(train_obj.shape[-1])
    ]
    model = ModelListGP(*models)
    mll = SumMarginalLogLikelihood(model.likelihood, model)
    return mll, model
```

**Reply by A-ep93** [reply]
Thanks for your response @saitcakmak. In the previous replies of this post, it has been mentioned that similar to a single output, for multi-output also, we need to use ExactMarginalLogLikelihood. However, in the code you sent, you used SumMarginalLogLikelihood. Can you please let me know the reason for this?

**Reply by Balandat** [reply]
> I can't really think of any particular reason other than that using a SingleTaskGP leads to fewer lines of code and keeps the tutorial simple

One other reason for using `SingleTaskGP` is that since it uses pytorch batch evaluation under the hood (rather than sequential evaluation of two models) it is usually faster to do predictions and thus optimize the acquisition function.
